sasl_transport.go

```diff
@@ -10,6 +10,43 @@ import (
 	"io"
 )
 
+type SaslConfig interface {
+	ConfigMap() map[string]string
+}
+
+type PlainSaslConfig struct {
+	Username string
+	Password string
+}
+
+func NewPlainSaslConfig(username, password string) *PlainSaslConfig {
+	return &PlainSaslConfig{
+		Username: username,
+		Password: password,
+	}
+}
+
+func (pc *PlainSaslConfig) ConfigMap() map[string]string {
+	return map[string]string{
+		"mechanismName": "PLAIN",
+		"username":      pc.Username,
+		"password":      pc.Password,
+	}
+}
+
+type GSSAPISaslConfig struct{}
+
+func NewGSSAPISaslConfig() *GSSAPISaslConfig {
+	return &GSSAPISaslConfig{}
+}
+
+func (gc *GSSAPISaslConfig) ConfigMap() map[string]string {
+	return map[string]string{
+		"mechanismName": "GSSAPI",
+		"service":       "impala",
+	}
+}
```

Comment:
Actually, we can even remove that and just have options that will generate a configuration map

Will be something like this: 

```go
type Options struct {
 	PollIntervalSeconds float64
 	BatchSize           int64
 	ConnectionTimeout   time.Duration
 	SaslTransportConfig map[string]string
 }

func WithPlainSaslTransport(username, password string) Option {
 	return func(o *Options) {
 		o.SaslTransportConfig = map[string]string{
  		         "mechanismName": "PLAIN",
  		         "username":      username,
  		         "password":      password,
  	        }
 	}
 }

 func WithGSSAPISaslTransport() Option {
 	return func(o *Options) {
 		o.SaslTransportConfig = map[string]string{
  		        "mechanismName": "GSSAPI",
  		        "service":       "impala",
  	        }
 	}
 }
```

Reply:
That's really pretty, shows the power of the pattern.